spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java

```diff
@@ -407,6 +428,216 @@ public void close() throws IOException {
+
+      InternalRow partition = meta.getStruct(partitionOrdinal, deletePartitionRowWrapper.size());
+      StructProjection partitionProjection = deletePartitionProjections.get(specId);
+      partitionProjection.wrap(deletePartitionRowWrapper.wrap(partition));
+
+      String file = id.getString(fileOrdinal);
+      long position = id.getLong(positionOrdinal);
+      delegate.delete(file, position, spec, partitionProjection);
+    }
+
+    @Override
+    public void update(InternalRow meta, InternalRow id, InternalRow row) throws IOException {
+      delete(meta, id);
+      delegate.update(row, dataSpec, null);
+    }
+
+    @Override
+    public void insert(InternalRow row) throws IOException {
+      delegate.insert(row, dataSpec, null);
+    }
+
+    @Override
+    public WriterCommitMessage commit() throws IOException {
+      close();
+
+      WriteResult result = delegate.result();
+      return new DeltaTaskCommit(result);
+    }
+
+    @Override
+    public void abort() throws IOException {
+      close();
+
+      WriteResult result = delegate.result();
+      cleanFiles(io, Arrays.asList(result.dataFiles()));
+      cleanFiles(io, Arrays.asList(result.deleteFiles()));
+    }
+
+    @Override
+    public void close() throws IOException {
+      if (!closed) {
+        delegate.close();
+        this.closed = true;
+      }
+    }
+  }
+
+  private static class PartitionedDeltaWriter extends BaseDeltaWriter {
+    private final PositionDeltaWriter<InternalRow> delegate;
+    private final FileIO io;
+    private final PartitionSpec dataSpec;
+    private final Map<Integer, PartitionSpec> specs;
+    private final InternalRowWrapper deletePartitionRowWrapper;
+    private final Map<Integer, StructProjection> deletePartitionProjections;
+    private final PartitionKey dataPartitionKey;
+    private final InternalRowWrapper internalRowDataWrapper;
+    private final int specIdOrdinal;
+    private final int partitionOrdinal;
+    private final int fileOrdinal;
+    private final int positionOrdinal;
+
+    private boolean closed = false;
+
+    PartitionedDeltaWriter(Table table, SparkFileWriterFactory writerFactory,
+                           OutputFileFactory dataFileFactory, OutputFileFactory deleteFileFactory,
+                           Context context) {
+
+      Preconditions.checkArgument(table.spec().isPartitioned(), "Data spec must be partitioned");
+
+      PartitioningWriter<InternalRow, DataWriteResult> insertWriter;
+      if (context.fanoutWriterEnabled()) {
+        insertWriter = new FanoutDataWriter<>(
+            writerFactory, dataFileFactory, table.io(),
+            context.dataFileFormat(), context.targetDataFileSize());
+      } else {
+        insertWriter = new ClusteredDataWriter<>(
+            writerFactory, dataFileFactory, table.io(),
+            context.dataFileFormat(), context.targetDataFileSize());
+      }
+
+      // always use a separate fanout data writer for updates as they may be out of order
```

Comment:
Got it. This is what I was looking for.

Reply:
If I remember correctly, this is because we sort by partition, file, and pos, but the correct output order depends on the new data. So we always delete with clustering, but the updated rows may not still conform to the original clustering. Right?

Reply:
Yeah, happens when we update a column that is used in the spec and the updated record moves to a different partition.